quartz/depgraph.ts

```diff
@@ -0,0 +1,189 @@
+export default class DepGraph<T> {
+  // node: incoming and outgoing edges
+  _graph = new Map<T, { incoming: Set<T>; outgoing: Set<T> }>()
+
+  constructor() {
+    this._graph = new Map()
+  }
+
+  export(): Object {
+    return {
+      nodes: this.nodes,
+      edges: this.edges,
+    }
+  }
+
+  toString(): string {
+    return JSON.stringify(this.export(), null, 2)
+  }
+
+  // BASIC GRAPH OPERATIONS
+
+  get nodes(): T[] {
+    return Array.from(this._graph.keys())
+  }
+
+  get edges(): [T, T][] {
+    let edges: [T, T][] = []
+    for (const [node, { outgoing }] of this._graph.entries()) {
+      outgoing.forEach((target) => {
```

Comment:
weird to mix `for` and `forEach` like this, let's go with one or the other

Reply:
entries() is an iterator so it doesn't have forEach, used for